syntax = 'proto3';

package r2t2.protobuf;

message InvocationPayload {
    string storage_backend = 1;
    string coordinator = 2;
    int32 samples_per_pixel = 3;
    int32 max_path_depth = 4;
    uint32 bagging_delay = 5;
    float ray_log_rate = 6;
    float bag_log_rate = 7;
    bool directional_treelets = 8;
    repeated string memcached_servers = 9;
}

message SceneObject {
    uint32 type = 1;
    uint64 id = 2;
    string alt_name = 3;
};


message SetupAccumulator
{
    string job_id = 1;
    string storage_backend = 2;
    uint32 width = 3;
    uint32 height = 4;
    uint32 tile_id = 5;
    uint32 tile_count = 6;
    repeated SceneObject scene_objects = 7;
}

message Hey {
    uint64 worker_id = 1;
    string job_id = 2;
}

message GetObjects {
    repeated SceneObject objects = 1;
}

message GenerateRays {
    int32 x0 = 1;
    int32 y0 = 2;
    int32 x1 = 3;
    int32 y1 = 4;
}

// Ray Bags

message RayBagInfo {
    bool tracked = 1;
    uint64 worker_id = 2;
    uint32 treelet_id = 3;
    uint64 bag_id = 4;
    uint64 ray_count = 5;
    uint64 bag_size = 6;
    bool sample_bag = 7;
}

message RayBags {
    uint64 rays_generated = 1;
    uint64 rays_terminated = 2;

    repeated RayBagInfo items = 3;
}

message WorkerStats {
    uint64 finished_paths = 1;
    double cpu_usage = 2;
}

// Benchmarking

message BenchmarkRequest {
    uint64 worker_id = 1;
    string storage_uri = 2;
    uint64 bag_size = 3;
    uint32 threads = 4;
    uint32 duration = 5;
    uint32 send = 6;
    uint32 recv = 7;
    repeated string memcached_servers = 8;
}

message BenchmarkResponse {
    string output = 1;
    string error = 2;
    string command = 3;
    int32 retcode = 4;
}

message HistogramUInt64 {
    uint64 width = 1;
    uint64 lower_bound = 2;
    uint64 upper_bound = 3;

    repeated uint64 bins = 4;

    uint64 count = 5;
    uint64 min = 6;
    uint64 max = 7;
    uint64 sum = 8;
    uint64 squares_sum = 9;
}

message JobStatus {
    string url_prefix = 1;
    uint64 next_sample_index = 2;
    repeated string sample_bags = 3;
}

message CategoryInfo {
    string title = 1;
    uint64 total_us = 2;
    double percentage = 3;
    uint64 max_us = 4;
    uint64 count = 5;
}

message TimeLog {
    uint64 total = 1;
    repeated CategoryInfo categories = 2;
    CategoryInfo unaccounted = 3;
}

message AccumulatedStats {
    message ItemInt64     { string title = 1; int64  value = 2; }
    message ItemFloat     { string title = 1; double value = 2; }
    message ItemInt64Pair { string title = 1; int64  first = 2; int64 second = 3; }

    repeated ItemInt64 counters = 1;
    repeated ItemInt64 memory_counters = 2;

    repeated ItemInt64 int_distribution_sums = 3;
    repeated ItemInt64 int_distribution_counts = 4;
    repeated ItemInt64 int_distribution_mins = 5;
    repeated ItemInt64 int_distribution_maxs = 6;

    repeated ItemFloat float_distribution_sums = 7;
    repeated ItemInt64 float_distribution_counts = 8;
    repeated ItemFloat float_distribution_mins = 9;
    repeated ItemFloat float_distribution_maxs = 10;

    repeated ItemInt64Pair percentages = 11;
    repeated ItemInt64Pair ratios = 12;
}

// Job Summary

message Point2 {
    int32 x = 1;
    int32 y = 2;
}

message JobSummary {
    // Job Info
    string job_id = 1;
    uint32 num_lambdas = 2;
    uint32 num_generators = 3;
    uint32 num_accumulators = 27;
    uint32 treelet_count = 4;
    Point2 output_size = 5;
    uint32 spp = 6;
    Point2 tile_size = 7;
    uint32 max_depth = 8;
    uint32 bagging_delay = 9;
    uint32 memcached_servers = 10;
    string storage_backend = 11;
    map<string, string> alt_scene_objects = 25;

    double total_time = 12;
    double generation_time = 13;
    double initialization_time = 14;
    double tracing_time = 15;
    uint64 total_paths = 16;
    uint64 finished_paths = 17;
    uint64 finished_rays = 18;
    uint64 num_enqueues = 19;
    double ray_throughput = 20;
    uint64 total_upload = 21;
    uint64 total_download = 22;
    uint64 total_samples = 23;
    double estimated_cost = 24;

    AccumulatedStats pbrt_stats = 26;
}
